components/client/src/main/java/com/hotels/styx/client/StyxBackendServiceClient.kt

```diff
@@ -0,0 +1,307 @@
+                    retry(updatedRequest, retryContext, newPreviousOrigins, attempt + 1, cause, context)
+                }
+                .map { response: LiveHttpResponse -> addOriginId(host.id(), response) }
+        } else {
+            val retryContext = RetryPolicyContext(id, attempt + 1, null, request, previousOrigins)
+            retry(request, retryContext, previousOrigins, attempt + 1, NoAvailableHostsException(id), context)
+        }
+    }
+
+    private fun addOriginId(originId: Id, response: LiveHttpResponse): LiveHttpResponse {
+        return response.newBuilder()
+            .header(originIdHeader, originId)
+            .build()
+    }
+
+    private fun retry(
+        request: LiveHttpRequest,
+        retryContext: RetryPolicyContext,
+        previousOrigins: List<RemoteHost>,
+        attempt: Int,
+        cause: Throwable,
+        context: HttpInterceptor.Context
+    ): Flux<LiveHttpResponse> {
+        val lbContext: LoadBalancer.Preferences = object : LoadBalancer.Preferences {
+            override fun preferredOrigins(): Optional<String> {
+                return Optional.empty()
+            }
+
+            override fun avoidOrigins(): List<Origin> {
+                return previousOrigins.map { it.origin() }
+            }
+        }
+        return if (retryPolicy.evaluate(retryContext, loadBalancer, lbContext).shouldRetry()) {
+            Flux.from(sendRequest(request, previousOrigins, attempt, context))
+        } else {
+            Flux.error(cause)
+        }
+    }
+
+    private class RetryPolicyContext(
+        private val appId: Id,
+        private val retryCount: Int,
+        private val lastException: Throwable?,
+        private val request: LiveHttpRequest,
+        private val previouslyUsedOrigins: Iterable<RemoteHost>
+    ) : RetryPolicy.Context {
+        override fun appId(): Id = appId
+
+
+        override fun currentRetryCount(): Int = retryCount
+
+
+        override fun lastException(): Optional<Throwable> = Optional.ofNullable(lastException)
+
+
+        override fun currentRequest(): LiveHttpRequest = request
+
+
+        override fun previousOrigins(): Iterable<RemoteHost> = previouslyUsedOrigins
+
+        override fun toString(): String =
+            StringBuilder()
+                .append("appId", appId)
+                .append(", retryCount", retryCount)
+                .append(", lastException", lastException)
+                .append(", request", request.url())
+                .append(", previouslyUsedOrigins", previouslyUsedOrigins)
+                .toString()
+
+        fun hosts(): String = hosts(previouslyUsedOrigins)
+
+        companion object {
+            private fun hosts(origins: Iterable<RemoteHost>): String {
+                return origins.asSequence().map { it.origin().hostAndPortString() }.joinToString(", ")
+            }
+        }
+    }
+
+    private fun logError(request: LiveHttpRequest, throwable: Throwable) {
```

Comment:
some of these are still just single expressions that don't require a method body. do review.